### PR TITLE
Feature/cache permissions

### PR DESCRIPTION
**Pull request by Maria-Bordunova**

### CODE REVIEW

File: sdk/src/main/java/com/qonversion/android/sdk/QProductCenterManager.kt [hunk 1]
```diff
@@ -407,32 +420,43 @@ class QProductCenterManager internal constructor(
...
     private fun loadStoreProductsIfPossible(
-        launchResult: QLaunchResult,
-        onLoadCompleted: ((products: List<SkuDetails>) -> Unit)?,
-        onLoadFailed: ((error: BillingError) -> Unit)?
+        onLoadCompleted: ((products: List<SkuDetails>) -> Unit)? = null,
+        onLoadFailed: ((error: BillingError) -> Unit)? = null
     ) {
+        if(isProductsLoaded && !isProductsLoadingFailed){
+            executeProductsBlocks()
+            return
```
Comment (Maria-Bordunova) [inline sdk/src/main/java/com/qonversion/android/sdk/QProductCenterManager.kt:446]: Is it necessary to call executeProductsBlocks() here?

File: sdk/src/main/java/com/qonversion/android/sdk/QProductCenterManager.kt [hunk 2]
```diff
@@ -407,32 +420,43 @@ class QProductCenterManager internal constructor(
...
+        }
+
+        val launchResult: QLaunchResult = getAvailableLaunchResult() ?: run{
+            executeProductsBlocks()
+            return
+        }
```
Comment (Maria-Bordunova) [inline sdk/src/main/java/com/qonversion/android/sdk/QProductCenterManager.kt:452]: The same question

File: sdk/src/main/java/com/qonversion/android/sdk/QProductCenterManager.kt [hunk 3]
```diff
@@ -513,23 +538,108 @@ class QProductCenterManager internal constructor(
...
+            it.onError(error)
+        }
+    }
+
+    private fun productForID(id: String?): QProduct? = getAvailableLaunchResult()?.products?.get(id)
+
+
+    private fun getAvailableLaunchResult(): QLaunchResult? =
+        this@QProductCenterManager.launchResult ?: launchResultCache.getLaunchResult()
+
```
Comment (Maria-Bordunova) [inline sdk/src/main/java/com/qonversion/android/sdk/QProductCenterManager.kt:641]: The method is used for loadStoreProductsIfPossible() and productForID(). I use getLaunchResult() regardless of cache timestamp. Or we should use getActualLaunchResult() for the first version

File: sdk/src/main/java/com/qonversion/android/sdk/QProductCenterManager.kt [hunk 4]
```diff
@@ -513,33 +556,118 @@ class QProductCenterManager internal constructor(
...
+    private fun handleLoadStateForProducts(onCompleted: () -> Unit) {
+        when (loadProductsState) {
+            Loaded -> onCompleted()
+            Failed -> loadStoreProductsIfPossible()
+            else -> Unit
+        }
+    }
+
+    private fun handleSuccessProducts(
```
Comment (suriksarkisyan) [inline sdk/src/main/java/com/qonversion/android/sdk/QProductCenterManager.kt:595]: Looks like this function used only once. Maybe you should remove and use this code block directly

File: sdk/src/main/java/com/qonversion/android/sdk/QProductCenterManager.kt [hunk 5]
```diff
@@ -513,33 +556,118 @@ class QProductCenterManager internal constructor(
...
+        permissions: Map<String, QPermission>
+    ) {
+        callbacks.forEach {
+            it.onSuccess(permissions)
         }
     }
 
-    private fun productForID(id: String?): QProduct? {
-        return launchResult?.products?.get(id)
+    private fun handleFailurePermissions(
```
Comment (suriksarkisyan) [inline sdk/src/main/java/com/qonversion/android/sdk/QProductCenterManager.kt:648]: Same about single usage

File: sdk/src/main/java/com/qonversion/android/sdk/QProductCenterManager.kt [hunk 6]
```diff
@@ -513,33 +556,118 @@ class QProductCenterManager internal constructor(
...
+                    val cachedLaunchResult = launchResultCache.getActualLaunchResult()
+
+                    cachedLaunchResult?.let {
+                        onSuccess(it.permissions)
+                    } ?: onError(error)
+                }
+            })
+    }
+
+    private fun handleSuccessPermissions(
```
Comment (suriksarkisyan) [inline sdk/src/main/java/com/qonversion/android/sdk/QProductCenterManager.kt:639]: Same about single usage

File: sdk/src/main/java/com/qonversion/android/sdk/QProductCenterManager.kt [hunk 7]
```diff
@@ -513,33 +556,118 @@ class QProductCenterManager internal constructor(
...
+    private fun handleFailurePermissions(
+        callbacks: List<QonversionPermissionsCallback>,
+        error: QonversionError
+    ) {
+        callbacks.forEach {
+            it.onError(error)
+        }
     }
 
+    private fun productForID(id: String?, launchResult: QLaunchResult): QProduct? =
```
Comment (suriksarkisyan) [inline sdk/src/main/java/com/qonversion/android/sdk/QProductCenterManager.kt:657]: Looks like this one-line function should be removed

File: sdk/src/main/java/com/qonversion/android/sdk/ad/LoadProductsState.kt [hunk 8]
```diff
@@ -0,0 +1,8 @@
+package com.qonversion.android.sdk.ad
+
+enum class LoadProductsState {
```
Comment (suriksarkisyan) [inline sdk/src/main/java/com/qonversion/android/sdk/ad/LoadProductsState.kt:3]:
```suggestion
enum class LoadStoreProductsState {
```

File: sdk/src/main/java/com/qonversion/android/sdk/storage/Cache.kt [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+import com.squareup.moshi.JsonAdapter
+
+
+interface Cache {
+
+    fun putInt(key: String, value: Int)
+    /**
+     * @param defValue is returned if the Int preference for key does not exist
+     */
+    fun getInt(key: String, defValue: Int = 0): Int
```
Comment (suriksarkisyan) [inline sdk/src/main/java/com/qonversion/android/sdk/storage/Cache.kt:12]: I think will be better to remove default value

**Review by suriksarkisyan — APPROVED** [review]

**Pull request merged**